File: sdk-typescript/src/utils/signature-utils.ts
import * as crypto from 'crypto';

/**
 * Derives the actual HMAC key from the provided secret using SHA256 hashing and string concatenation.
 *
 * @param initialSecret The initial secret.
 * @returns The SHA256-derived HMAC key as an uppercase hexadecimal string.
 */
function deriveHmacKey(initialSecret: string): string {
    const concatenatedString = `${initialSecret}:reference_token`;

    const hash = crypto.createHash('sha256')
                        .update(Buffer.from(concatenatedString, 'utf8'))
                        .digest();

    return hash.toString('hex').toUpperCase();
}

/**
 * Computes the HMAC-SHA1 signature for a given webhook payload and secret.
 *
 * @param secret The webhook secret (your `referenceToken`).
 * @param payload The raw JSON string of the webhook body.
 * @returns The Base64-encoded HMAC-SHA1 signature.
 */
export function computeWebhookSignature(secret: string, payload: string): string {
    if (!secret) {
        throw new Error("Webhook secret cannot be empty.");
    }
    const hmacKey = deriveHmacKey(secret);
    const hmac = crypto.createHmac('sha1', Buffer.from(hmacKey , 'utf8'));
    hmac.update(Buffer.from(payload, 'utf8'));
    return hmac.digest('base64');
}

/**
 * Verifies the authenticity of a Pearl webhook payload using its signature.
 *
 * @param receivedSignature The signature received in the 'X-Pearl-API-Signature' header.
 * @param webhookPayloadJsonString The raw JSON string of the webhook request body.
 * @param webhookSecret The shared secret.
 * @returns boolean indicating if the signature is valid.
 * @throws Error if the signature is invalid for security reasons (prevents silent failures),
 * or if required parameters are missing.
 */
export function verifyWebhookSignature(
    receivedSignature: string,
    webhookPayloadJsonString: string,
    webhookSecret: string
): boolean {
    if (!receivedSignature || !webhookPayloadJsonString || !webhookSecret) {
        throw new Error("Missing required parameters for webhook signature verification.");
    }

    const computedSignature = computeWebhookSignature(webhookSecret, webhookPayloadJsonString);

    const receivedSigBuffer = Buffer.from(receivedSignature, 'base64');
    const computedSigBuffer = Buffer.from(computedSignature, 'base64');

    if (receivedSigBuffer.length !== computedSigBuffer.length) {
        return false;
    }

    return crypto.timingSafeEqual(receivedSigBuffer, computedSigBuffer);
}

File: sdk-python/pearl_sdk/utils/signature_utils.py
"""
Utilities for webhook signature computation and verification.
"""

import base64
import hashlib
import hmac


def _derive_hmac_key(initial_secret: str) -> str:
    """
    Derives the actual HMAC key from the provided secret using SHA256 hashing and string concatenation.
    
    Args:
        initial_secret: The initial secret.
        
    Returns:
        The SHA256-derived HMAC key as an uppercase hexadecimal string.
    """
    concatenated_string = f"{initial_secret}:reference_token"
    hash_bytes = hashlib.sha256(concatenated_string.encode('utf-8')).digest()
    return hash_bytes.hex().upper()


def compute_webhook_signature(secret: str, payload: str) -> str:
    """
    Computes the HMAC-SHA1 signature for a given webhook payload and secret.
    
    Args:
        secret: The webhook secret (your `referenceToken`).
        payload: The raw JSON string of the webhook body.
        
    Returns:
        The Base64-encoded HMAC-SHA1 signature.
        
    Raises:
        ValueError: If the webhook secret is empty.
    """
    if not secret:
        raise ValueError("Webhook secret cannot be empty.")
    
    hmac_key = _derive_hmac_key(secret)
    hmac_digest = hmac.new(
        hmac_key.encode('utf-8'),
        payload.encode('utf-8'),
        hashlib.sha1
    ).digest()
    
    return base64.b64encode(hmac_digest).decode('utf-8')


def verify_webhook_signature(
    received_signature: str,
    webhook_payload_json_string: str,
    webhook_secret: str
) -> bool:
    """
    Verifies the authenticity of a Pearl webhook payload using its signature.
    
    Args:
        received_signature: The signature received in the 'X-Pearl-API-Signature' header.
        webhook_payload_json_string: The raw JSON string of the webhook request body.
        webhook_secret: The shared secret.
        
    Returns:
        boolean indicating if the signature is valid.
        
    Raises:
        ValueError: If required parameters are missing.
    """
    if not received_signature or not webhook_payload_json_string or not webhook_secret:
        raise ValueError("Missing required parameters for webhook signature verification.")

    computed_signature = compute_webhook_signature(webhook_secret, webhook_payload_json_string)
    
    try:
        received_sig_bytes = base64.b64decode(received_signature)
        computed_sig_bytes = base64.b64decode(computed_signature)
    except Exception:
        return False
    
    if len(received_sig_bytes) != len(computed_sig_bytes):
        return False
    
    return hmac.compare_digest(received_sig_bytes, computed_sig_bytes)
